File: leetcode/DynamicProgram/leetcode-213/213.go
package dp

func rob(nums []int) int {
	if len(nums) == 0 {
		return 0
	} else if len(nums) == 1 {
		return nums[0]
	} else if len(nums) == 2 {
		return max(nums[0], nums[1])
	}

	prepre, pre := nums[0], max(nums[0], nums[1])
	res := 0

	for i := 2; i < len(nums)-1; i++ {
		tmp := pre
		pre = max(prepre+nums[i], pre)
		prepre = tmp
	}
	res = max(res, pre)

	prepre, pre = nums[1], max(nums[1], nums[2])
	for i := 3; i < len(nums); i++ {
		tmp := pre
		pre = max(prepre+nums[i], pre)
		prepre = tmp
	}
	res = max(res, pre)

	return res
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: leetcode/DynamicProgram/leetcode-213/213.java
class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        if (nums.length == 1) {
            return nums[0];
        }

        if (nums.length == 2) {
            return Math.max(nums[0], nums[1]);
        }

        int prepre = nums[0];
        int pre = Math.max(nums[0], nums[1]);

        int res = 0;

        for (int i = 2; i < nums.length - 1; i++) {
            int tmp = pre;
            pre = Math.max(prepre + nums[i], pre);
            prepre = tmp;
        }
        res = Math.max(res, pre);

        prepre = nums[1];
        pre = Math.max(nums[1], nums[2]);
        for (int i = 3; i < nums.length; i++) {
            int tmp = pre;
            pre = Math.max(prepre + nums[i], pre);
            prepre = tmp;
        }
        res = Math.max(res, pre);

        return res;
    }
}